config/types/files.go

```diff
@@ -63,9 +63,11 @@ type File struct {
 }
 
 type FileContents struct {
-	Remote Remote `yaml:"remote"`
-	Inline string `yaml:"inline"`
-	Local  string `yaml:"local"`
+	Base64      *bool  `yaml:"base64"`
```

Comment:
I would rather we leave this up to the YAML parser to interpret this via type tagging. See http://yaml.org/type/binary.html for the binary tag.

Reply:
Hi @crawford,

Are you suggesting that I modify `Inline` to be a different type then? Or are you suggesting this as a CT example:

```yaml
storage:
  files:
    - path: /etc/helloworld.txt
      filesystem: root
      mode: 0644
      contents:
        compression: gzip
        inline: !!binary "\
          H4sIAB/jFlsAA/NIzcnJ11Eozy/KSdHjAgDXu838DgAAAA=="
```

Reply:
I'm suggesting the CLC example be changed to what you show:

```yaml
storage:
  files:
    - path: /etc/helloworld.txt
      filesystem: root
      mode: 0644
      contents:
        compression: gzip
        inline: !binary |-
          H4sIAB/jFlsAA/NIzcnJ11Eozy/KSdHjAgDXu838DgAAAA==
```

Reply:
Okay. I do not know to detect that tag in Go. I will look into it.

Reply:
The yaml parser _should_ handle it automatically

Reply:
Hi @ajeddeloh,

Okay, but I still am not sure how I then detect that the string is base-64 encoded and thus set the data URL's `base64` attribute from that? Because *that* is the reason for the boolean in the first place. Without setting that, Ignition's URL parser won't know to base-64 decode the data string.

Ignition uses the Go package [`github.com/vincent-petithory/dataurl`](https://github.com/vincent-petithory/dataurl) to parse URLs for a possible data element. This package properly [detects](https://github.com/vincent-petithory/dataurl/blob/master/dataurl.go#L143-L151) the base64 encoding where Go's doesn't. Unless this is set, the package assumes the data value is URL-encoded. 

Do you see the change I made to the way the value is encoded if the `Base64` field is set to `true`?

```go
if file.Contents.Base64 != nil && *file.Contents.Base64 {
    newFile.Contents = ignTypes.FileContents{
        Source: (&url.URL{
            Scheme: "data",
            Opaque: ";base64," + file.Contents.Inline,
        }).String(),
    }
}
```

The `Opaque` field adds that `;base64,` section and the contents are no longer URL encoded.

Reply:
When you tag the type as "binary", the YAML parser will do the conversion for you. By the time the `FileContents` structure is populated, the contents will already be decoded. This mechanism is separate from the dataurl's encoding (e.g. `data://;base64,<data>`). The resulting dataurl won't be base64 encoded, but will instead be URL encoded.

In the future, we could change the dataurl logic to base64 encode contents under certain circumstances (i.e. when the URL encoded size is larger than base64 encoding).

Reply:
Hi @crawford,

I see what you're saying. Hmm. Interesting.

Reply:
Hi @crawford,

I take that back, I don't see. Are you saying that the gzip'd data would be URL encoded? Is the gzip'd data already base64 encoded? If so, what's the point of URL encoding anything? I must be missing something. I do not see how the binary notation works with the idea of supplying compressed, gzip'd data as local contents.

FWIW, I *am* aware of how it works currently with the data URL per the way `contents.local` and `--files-dir` works.

I suppose what *I'm* trying to supply is a way to provide gzip'd, base64-encoded data that Ignition will know how to decode and decompress before writing to a file.

Reply:
Given the following config:

```yaml
storage:
  files:
    - path: /etc/helloworld.txt
      filesystem: root
      mode: 0644
      contents:
        compression: gzip
        inline: !binary |-
          H4sIAB/jFlsAA/NIzcnJ11Eozy/KSdHjAgDXu838DgAAAA==
```

The YAML library would parse this into the following Go struct:

```go
FileContents {
  Compression: "gzip",
  Inline: "\x1f\x8b\x08\x00\x1f\xe3\x16\x5b\x00\x03\xf3\x48\xcd\xc9\xc9\xd7\x51\x28\xcf\x2f\xca\x49\xd1\xe3\x02\x00\xd7\xbb\xcd\xfc\x0e",
}
```

The inline string is the gzipped string: Hello, world.

CT would then emit the following JSON:

```json
{
  "ignition": {
    "config": {},
    "security": {
      "tls": {}
    },
    "timeouts": {},
    "version": "2.2.0"
  },
  "networkd": {},
  "passwd": {},
  "storage": {
    "files": [
      {
        "filesystem": "root",
        "path": "/etc/helloworld.txt",
        "contents": {
          "source": "data:,H4sIAB%2FjFlsAA%2FNIzcnJ11Eozy%2FKSdHjAgDXu838DgAAAA%3D%3D",
          "compression": "gzip",
          "verification": {}
        },
        "mode": 420
      }
    ]
  },
  "systemd": {}
}
```

Ignition would (with some modification) decompress the dataurl, revealing "Hello, world.".

Reply:
Hi @crawford,

Now I actually do see. Thank you. I usually get it. But it also usually takes me a while. My mama says I'm smart in a special way. :)

Okay, so, sounds like I can just get rid of the `Base64` field and how I refactored the encoding logic. I just need @ajeddeloh to respond re: the `Compression` field. @crawford, do you have a preference? Please note I only maintained the previous field to maintain backwards-compatibility.

Reply:
@figo Please open an issue if you are having trouble.

Reply:
@crawford i am happy to do that if you could tell me where should the issue goes, it is ignition, could https://github.com/coreos/terraform-provider-ct/issues be a proper place? thanks

Reply:
@figo https://issues.coreos.com

Reply:
thanks, issue filed, https://github.com/coreos/bugs/issues/2458